oxidation/libparsec/crates/core/src/invite/claimer.rs

```diff
@@ -0,0 +1,488 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 (eventually AGPL-3.0) 2016-present Scille SAS
+
+use libparsec_client_connection::InvitedCmds;
+use libparsec_client_types::LocalDevice;
+use libparsec_crypto::{generate_nonce, HashDigest, PrivateKey, SecretKey, SigningKey};
+use libparsec_protocol::invited_cmds::v2::*;
+use libparsec_types::{
+    BackendOrganizationAddr, DeviceLabel, HumanHandle, InviteDeviceConfirmation, InviteDeviceData,
+    InviteUserConfirmation, InviteUserData, SASCode, UserID,
+};
+
+use crate::{InviteError, InviteResult};
+
+pub async fn claimer_retrieve_info(cmds: &InvitedCmds) -> InviteResult<invite_info::UserOrDevice> {
+    let rep = cmds.send(invite_info::Req).await?;
+
+    match rep {
+        invite_info::Rep::Ok(claimer) => Ok(claimer),
+        invite_info::Rep::UnknownStatus { unknown_status, .. } => Err(InviteError::Custom(
+            format!("Backend error during invitation retrieval: {unknown_status}"),
+        )),
+    }
+}
+
+#[derive(Debug)]
+struct BaseClaimInitialCtx {
+    greeter_user_id: UserID,
+    greeter_human_handle: Option<HumanHandle>,
+    cmds: InvitedCmds,
+}
+
+impl BaseClaimInitialCtx {
+    async fn do_wait_peer(self) -> InviteResult<BaseClaimInProgress1Ctx> {
+        let claimer_private_key = PrivateKey::generate();
+
+        let rep = self
+            .cmds
+            .send(invite_1_claimer_wait_peer::Req {
+                claimer_public_key: claimer_private_key.public_key(),
+            })
+            .await?;
+
+        let greeter_public_key = match rep {
+            invite_1_claimer_wait_peer::Rep::AlreadyDeleted => {
+                return Err(InviteError::AlreadyUsed)
+            }
+            invite_1_claimer_wait_peer::Rep::InvalidState => return Err(InviteError::PeerReset),
+            invite_1_claimer_wait_peer::Rep::NotFound => return Err(InviteError::NotFound),
+            invite_1_claimer_wait_peer::Rep::Ok { greeter_public_key } => greeter_public_key,
+            invite_1_claimer_wait_peer::Rep::UnknownStatus { unknown_status, .. } => {
+                return Err(InviteError::Custom(format!(
+                    "Backend error during step 1: {unknown_status}"
+                )))
+            }
+        };
```

Comment:
```suggestion
        let greeter_public_key = match rep {
            invite_1_claimer_wait_peer::Rep::AlreadyDeleted => Err(InviteError::AlreadyUsed),
            invite_1_claimer_wait_peer::Rep::InvalidState => Err(InviteError::PeerReset),
            invite_1_claimer_wait_peer::Rep::NotFound => Err(InviteError::NotFound),
            invite_1_claimer_wait_peer::Rep::Ok { greeter_public_key } => Ok(greeter_public_key),
            invite_1_claimer_wait_peer::Rep::UnknownStatus { unknown_status, .. } => {
                Err(InviteError::Custom(format!(
                    "Backend error during step 1: {unknown_status}"
                )))
            }
        }?;
```

Is using `?` better than multiple `return` :thinking: ?

Reply:
I don't know, it's shorter but it wraps when it's ok :man_shrugging:

Reply:
#4252

Reply:
I mark it as resolved then